### ISSUE TITLE
"Failed to clean and unmount volume" errors blocking deployments of Daemonsets

### ISSUE THREAD

**Issue by jroggeman** [kubernetes-sigs/secrets-store-csi-driver#936]
**What steps did you take and what happened:**
1. Deploy a `DaemonSet` that mounts a KeyVault CSI volume
2. Deploy an update to that Daemonset
3. There is a chance that pods will get in to a semi-permanent stuck state that blocks Daemonset deployments. Investigation, logs, and details below.

Note - I believe this could happen for any pods, but causes the most issues with Daemonsets, for two reasons:
- We have large clusters, so we have considerably more pod replicas for our DSs
- DSs don't allow surging, so if a pod won't delete, it blocks the daemonset from fully rolling out indefinitely.

**What did you expect to happen:**
All subsequent DaemonSet rollouts should be successful.

**Anything else you would like to add:**
As mentioned above, we're seeing blocked Daemonset rollouts because the DS pods using a CSI volume are failing to unmount - they get stuck in Terminating indefinitely. We're seeing logs from CSI's `nodeserver` along the following lines:

```
E0414 10:45:15.041479       1 nodeserver.go:263] "failed to clean and unmount target path" err="remove /var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/ingress-ssl-volume/mount: directory not empty" targetPath="/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/ingress-ssl-volume/mount"
```
After logging on to the node in question, we were able to confirm that the directory was indeed not empty.

We looked at other logs around the same time, and determined that there's a check done in [mount_helper_common](https://github.com/kubernetes/utils/blob/3a6ce19ff2f91f3dc49243f90f7834a0f1e4aaf3/mount/mount_helper_common.go#L50) that, on unmount, will check if a folder is "likely" not a mount point, and if it is not, just deletes it. The code for that check, [IsLikelyNotMountPoint](https://github.com/kubernetes/utils/blob/3a6ce19ff2f91f3dc49243f90f7834a0f1e4aaf3/mount/mount_linux.go#L259) (for linux, which we use), is running `os.stat` on the /mount dir and the parent dir containing it, and returns that it is likely not a mount point if they have the same device ID. We're seeing this log:

```
2022-04-15T19:16:51.577389123Z stderr F W0415 19:16:51.577244       1 mount_helper_common.go:129] Warning: "/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/csi-secrets/    mount" is not a mountpoint, deleting
```

I Did some manual validation on the nodes by running the `stat` command on those folders and checking the `Device ID` field. For healthy pods, the device ID does indeed differ:

```
root@<node-name>:/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/secrets-store-inline# ls
mount  vol_data.json
root@<node-name>:/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/secrets-store-inline# stat mount/
  File: mount/
  Size: 100             Blocks: 0          IO Block: 4096   directory
Device: 37h/55d Inode: 2           Links: 3
Access: (1777/drwxrwxrwt)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-04-15 19:51:43.013305632 +0000
Modify: 2022-04-15 19:51:43.225317550 +0000
Change: 2022-04-15 19:51:43.225317550 +0000
Birth: -

root@<node-name>:/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/secrets-store-inline# stat .
  File: .
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: 801h/2049d      Inode: 6451407     Links: 3
Access: (0750/drwxr-x---)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-04-15 23:02:11.823478004 +0000
Modify: 2022-04-15 19:51:43.013305632 +0000
Change: 2022-04-15 19:51:43.013305632 +0000
Birth: -
root@<node-name>:/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/secrets-store-inline#

```

BUT, on the pods that get in to a bad state, we actually see the mount folder (which still contains the mounted files we'd expect) with the _same_ device ID as the parent (i.e., an ID mapping to `/dev/sda1`):

```
root@<node-name>:/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/csi-secrets/mount# stat /var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/csi-secrets/mount
  File: /var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/csi-secrets/mount
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: 801h/2049d      Inode: 6451201     Links: 3
Access: (0750/drwxr-x---)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-04-15 21:44:43.750372862 +0000
Modify: 2022-04-15 06:51:35.041930329 +0000
Change: 2022-04-15 06:51:35.041930329 +0000
Birth: -


root@<node-name>:/var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/csi-secrets/mount# stat /var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/csi-secrets
  File: /var/lib/kubelet/pods/<pod-guid>/volumes/kubernetes.io~csi/csi-secrets
  Size: 4096            Blocks: 8          IO Block: 4096   directory
Device: 801h/2049d      Inode: 6451663     Links: 3
Access: (0750/drwxr-x---)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-04-15 20:23:00.977789806 +0000
Modify: 2022-04-15 20:22:13.318766226 +0000
Change: 2022-04-15 20:22:13.318766226 +0000
Birth: -

```
At that point, after bouncing back up to the mount helper util class, it believes it’s not a mount point, and calls os.remove() on the directory, which expects it to be empty (which it isn’t), causing the failure.

That's as far as I've been able to get so far. I'm hoping someone with some expertise in this area could help me debug the issue and either determine what we're doing wrong, or help with a workaround and eventual bugfix. I may be able to contribute the fix myself if it's small and scoped enough, depending on work bandwidth.

Few other quick notes:
- The condition seems to last for awhile. I haven't yet determined if it will autoheal - we see pods with this problem eventually go away, but it's possible/likely it's just new deployments, scaling, or node rebalancing.
- We can manually mitigate the issue by going in and doing a `delete --force` on the pod (probably leaving some dirty files around on the node), but it's not ideal, especially for normal deployment pipelines.
- I believe this could happen for any pods, but causes the most issues with Daemonsets, for two reasons:
  - We have large clusters, so we have considerably more pod replicas for our DSs
  - DSs don't allow surging, so if a pod won't delete, it blocks the daemonset from fully rolling out indefinitely.

**Which provider are you using:**
Azure Key vault Provider. Based on the logs, the errors don't appear to be specific to any files in that repo, hence why I"m starting here.

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): 1.0.0
- Kubernetes version: (use `kubectl version`):
  - kubectl client: 1.21.2
  - server: 1.22.6

**Comment by nilekhc** [comment]
After looking at the Secrets Store installation last week, we made couple changes.

1. Moved Secrets Store CSI into kube-system namespace rather than it's own.
2. Upgraded to latest version

After this we are not observing any failures so far. @jroggeman will try to repro this and we will continue to monitor.

**Comment by aramase** [comment]
The issue is mitigated with #964 that's included in [v1.2.0](https://github.com/kubernetes-sigs/secrets-store-csi-driver/releases/tag/v1.2.0) release. Please feel free to reopen if you have any questions or still have issues with the newer release.

/close
